rust/connlib/tunnel/src/sockets.rs

```diff
@@ -254,28 +262,40 @@ impl Socket {
         }
     }
 
-    fn poll_send_ready(&self, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
-        self.socket.poll_send_ready(cx)
+    fn poll_flush(&mut self, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        loop {
+            match self.socket.try_io(Interest::WRITABLE, || {
+                self.state
+                    .send((&self.socket).into(), &self.buffered_transmits)
+            }) {
+                Ok(0) => break,
+                Err(e) if e.kind() == io::ErrorKind::WouldBlock => break,
+                Err(e) => return Poll::Ready(Err(e)),
+
+                Ok(num_sent) => {
+                    self.buffered_transmits.drain(..num_sent);
+
+                    // I am not sure if we'd ever send less than what is in `buffered_transmits`.
+                    // loop once more to be sure we `break` on either an empty buffer or on `WouldBlock`.
+                }
+            };
+        }
+
+        // Ensure we are ready to send more data.
+        ready!(self.socket.poll_send_ready(cx)?);
+
+        assert!(
+            self.buffered_transmits.is_empty(),
+            "buffer must be empty if we are ready to send more data"
+        );
+
+        Poll::Ready(Ok(()))
     }
 
-    fn try_send_to(
-        &self,
-        local: Option<SocketAddr>,
-        dest: SocketAddr,
-        buf: &[u8],
-    ) -> io::Result<usize> {
-        tracing::trace!(target: "wire", to = "network", src = ?local, dst = %dest, num_bytes = %buf.len());
-
-        self.state.send(
-            (&self.socket).into(),
-            &[quinn_udp::Transmit {
-                destination: dest,
-                ecn: None,
-                contents: Bytes::copy_from_slice(buf),
-                segment_size: None,
-                src_ip: local.map(|s| s.ip()),
-            }],
-        )
+    fn send(&mut self, transmit: quinn_udp::Transmit) {
+        tracing::trace!(target: "wire", to = "network", src = ?transmit.src_ip, dst = %transmit.destination, num_bytes = %transmit.contents.len());
+
+        self.buffered_transmits.push(transmit);
```

Comment:
```suggestion
        if self.buffered_transmits.length() > 1_000_000_000 {
            panic!("Too many buffered outgoing packets");
        }
        self.buffered_transmits.push(transmit);
```

Couldn't hurt, right?

Reply:
> Because we enforce further send-readiness before continuing, this buffer cannot grow unbounded.

I saw this in the original post but I can't see where it's enforced. Maybe it's in another file?

Reply:
The enforcement of send-readiness is the `poll_send_ready` at the end of `flush` and the `ready!` within `Io::poll`.

Reply:
I can add an assertion for the size, yeah.

Reply:
See [7baa658](https://github.com/firezone/firezone/pull/4168/commits/7baa6582fe1daaea0ac10562d080cae587ca55f5).